File: backend/src/routes/access.js
'use strict';

/**
 * access_api — Access Controller
 * Routes: share document, revoke access
 * Manages permissions and collaborator relationships.
 */

const express = require('express');
const docRepo = require('../repositories/documents');
const userRepo = require('../repositories/users');
const { requireAuth } = require('../middleware/auth');

const router = express.Router({ mergeParams: true });
router.use(requireAuth);

// POST /api/documents/:id/share
// Body: { email, role }  role ∈ 'viewer' | 'editor'
router.post('/', (req, res) => {
  const doc = docRepo.findById(req.params.id);
  if (!doc) return res.status(404).json({ error: 'Document not found' });
  if (doc.owner_id !== req.user.id) return res.status(403).json({ error: 'Only the owner can share' });

  const { email, role } = req.body;
  if (!email || !['viewer', 'editor'].includes(role)) {
    return res.status(400).json({ error: 'email and role (viewer|editor) are required' });
  }

  const target = userRepo.findByEmail(email);
  if (!target) return res.status(404).json({ error: 'User not found' });
  if (target.id === req.user.id) return res.status(400).json({ error: 'Cannot share with yourself' });

  userRepo.upsertPermission(doc.id, target.id, role);

  res.status(201).json({
    permission: {
      user: { id: target.id, username: target.username, email: target.email },
      role,
    },
  });
});

// DELETE /api/documents/:id/share/:userId
router.delete('/:userId', (req, res) => {
  const doc = docRepo.findById(req.params.id);
  if (!doc) return res.status(404).json({ error: 'Document not found' });
  if (doc.owner_id !== req.user.id) return res.status(403).json({ error: 'Only the owner can revoke access' });

  userRepo.deletePermission(doc.id, req.params.userId);
  res.json({ message: 'Access revoked' });
});

module.exports = router;

File: backend/app/routers/access.py
from __future__ import annotations

from fastapi import APIRouter, Depends, HTTPException, status

from .. import repository
from ..deps import get_connection, get_current_user
from ..schemas import DeleteDocumentResponse, ShareDocumentRequest, ShareDocumentResponse

router = APIRouter(prefix="/api/documents/{document_id}/share", tags=["sharing"])


@router.post(
    "",
    response_model=ShareDocumentResponse,
    status_code=status.HTTP_201_CREATED,
    summary="Share a document with a user",
    description="Owner-only route that grants or updates viewer/editor access for a target user identified by username or email.",
)
def share_document(
    document_id: int,
    payload: ShareDocumentRequest,
    current_user=Depends(get_current_user),
    connection=Depends(get_connection),
):
    document = repository.find_document(connection, document_id)
    if document is None:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Document not found")
    if document["owner_id"] != current_user["id"]:
        raise HTTPException(status_code=status.HTTP_403_FORBIDDEN, detail="Only the owner can share")

    identifier = (payload.identifier or payload.email or "").strip()
    target = repository.find_user_by_identifier(connection, identifier)
    if target is None:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="User not found")
    if target["id"] == current_user["id"]:
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail="Cannot share with yourself")

    permission = repository.upsert_permission(connection, document_id, target["id"], payload.role)
    return {"permission": permission}


@router.delete(
    "/{user_id}",
    response_model=DeleteDocumentResponse,
    summary="Revoke direct collaborator access",
    description="Owner-only route that removes an existing direct collaborator permission from the document.",
)
def revoke_access(
    document_id: int,
    user_id: int,
    current_user=Depends(get_current_user),
    connection=Depends(get_connection),
):
    document = repository.find_document(connection, document_id)
    if document is None:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Document not found")
    if document["owner_id"] != current_user["id"]:
        raise HTTPException(status_code=status.HTTP_403_FORBIDDEN, detail="Only the owner can revoke access")

    repository.revoke_permission(connection, document_id, user_id)
    return {"message": "Access revoked"}
